### PR TITLE
tracing: fix Datadog span name

### PR DESCRIPTION
**Pull request by dgoffredo**
### Commit Message
tracing: fix Datadog span name

### Additional Description
In a [series of pull requests][1] beginning last year, I modified Envoy's Datadog tracing extension to use a new library instead of OpenTracing.

Those changes included a bug.

[Previously][2], a Datadog span produced by Envoy had its "operation name" hard-coded to "envoy.proxy". The `operation_name` parameter to the OpenTracing adapter's `startSpan` then effectively became the span's "resource name," because of how operation name and resource name were managed by the OpenTracing-based Datadog tracing library.

My changes instead set the span's operation name to [the `operation_name` parameter of `startSpan`][3].

Those changes were released with envoy v1.26.0.

The result is that upgrading Envoy to v1.26 changed spans that used to be:
```
operation name: envoy.proxy
resource name: egress localhost:8080
```
into instead:
```
operation name: egress localhost:8080
resource name:
```

To remedy this, this pull request contains changes that restore the old behavior.

If these changes are accepted, I would appreciate if they would be included in the current minor release series as a bug fix patch.

### Risk Level
low

### Testing
See the modified unit test.

Also, I ran the `sources/extensions/tracers/datadog/demo` both with and without these changes. Verified that the produced span's "operation name" before these changes is not as desired. Verified that the produced span's "operation name" after these changes is as desired.

Desired: "Operation name" is "envoy.proxy", and "resource name" is the `operation_name` passed to `startSpan`.

### Docs Changes
none

### Release Notes
Recommended wording:

> Fixed a bug in the Datadog tracer where Datadog's "operation name" would contain what should be in the "resource name."

### Platform Specific Features
none

[1]: https://github.com/envoyproxy/envoy/pulls?q=is%3Apr+author%3Adgoffredo
[2]: https://github.com/envoyproxy/envoy/blob/05964eb722797d8a300a4c24410438382e472f36/source/extensions/tracers/datadog/datadog_tracer_impl.cc#L38
[3]: https://github.com/envoyproxy/envoy/blob/51964702956d64adcd1df6b8ea132e863fe78e74/source/extensions/tracers/datadog/tracer.cc#L88

### CODE REVIEW

**Review by zuercher — COMMENTED** [review]
Please update `changeslogs/current.yaml` with your release note. Otherwise this looks good.

**Review by zuercher — COMMENTED** [review]
Also, please add a test in `test/extensions/tracers/datadog`. I think you could probably check this in the existing `ExtractionSuccess` test in `tracer_test.cc`.

**Comment by dgoffredo** [comment]
Thanks, @zuercher, that was quick!

I'll update `changeslogs/current.yaml` like you said. My recent commit should have fixed the unit test.

**Comment by dgoffredo** [comment]
Ok, changelog updated.

Once CI accepts this (it might need a poke) and it's approved, I'd like to revisit the question of onto which release series this change will appear. I don't know the process for requesting a patch to the current release series.

File: changelogs/current.yaml [hunk 1]
```diff
@@ -156,6 +156,9 @@ bug_fixes:
 - area: redis
   change: |
     Fixed a bug where redis key with % in the key is failing with a validation error.
+- area: tracing
+  change: |
+    Fixed a bug in the Datadog tracer where Datadog's "operation name" would contain what should be in the "resource name."
```
Comment (zuercher) [inline changelogs/current.yaml:161]:
```suggestion
    Fixed a bug in the Datadog tracer where Datadog's "operation name" field would contain what should be in the "resource name" field.
```

We have a format checker that requires these to end with a period and it apparently doesn't consider this to be correct. That's caused the CI failure.

**Review by zuercher — DISMISSED** [review]
RBE is having a sad, so CI is busted but otherwise this looks good.

**Comment by dgoffredo** [comment]
I just pushed some changes that merge recent additions to `changelogs/current.yaml` with these. Looks like Envoy just had a release.

**Comment by dgoffredo** [comment]
I just pushed some changes that merge yet more additions to changelogs/current.yaml with these.

@zuercher Thanks again for your review. How do we move this PR forward?

**Review by zuercher — APPROVED** [review]

**Comment by zuercher** [comment]
We need the tests to pass. I restarted the TSAN test and hopefully it will pass.

**Comment by florianmutter** [comment]
Hey, I'm not familiar with envoy releases. Can I see somewhere in which version this fix is or will be available? We wanted to update to Istio 1.19 but the change in span names does block us.

**Comment by phlax** [comment]
if you look at the commit you can see which branches its on

https://github.com/envoyproxy/envoy/commit/c3646f994e0296ed44ecac2869fa65e7f58bf986

currently its only on `main` which it means it hasnt been released anywhere but will be a part of 1.28+

@florianmutter if you want to raise backports for other release branches (1.24-7) i would be happy to review - it would then be included in any backported branches when they release, which is probably quite soon

**Comment by florianmutter** [comment]
I created the first backport PR for 1.26 here: https://github.com/envoyproxy/envoy/pull/30182

**Comment by hiddewie** [comment]
I tested this fix with Envoy 1.27.2, as part of the resolution of https://github.com/envoyproxy/envoy/issues/30235. The operation name has been fixed from `ingress` to `envoy.proxy`, but the resource name is still not taken from the tag `resource.name` like it was in Envoy versions 1.26.3 and earlier.

**Comment by dgoffredo** [comment]
@hiddewie Yes, the handling of special tag names was removed in the new Datadog tracing library because it is no longer based on OpenTracing.

However, as far as I can see, "resource.name" is not part of OpenTracing's [semantic conventions][1], and so removing it was a change in _Datadog_ behavior.

I can either restore the special meaning of the "resource.name" tag as another patch, or find another solution.

Out of curiosity, what is your particular use case for setting the "resource.name"?

[1]: https://github.com/opentracing/specification/blob/master/semantic_conventions.md

**Comment by hiddewie** [comment]
OK, thanks, interesting.

> Out of curiosity, what is your particular use case for setting the "resource.name"?

We set an internal header in the Envoy configuration responsible for routing the requests (there is Lua involved), to a functional resource name (basically sets of similar paths that have a functional name). This internal header is used in the `custom_tags` tracing configuration to set the `resource.name` tag. The resource name is used in the Datadog UI to provide summarized information about similar resources, like request count, latency information and error rate. This information per functional resource gives valuable insights into a subset of all proxied requests.

When there is only a single `ingress` resource present as part of the `envoy.proxy` operation, these summaries per functional resource name are not generated by Datadog.

**Comment by dgoffredo** [comment]
> We set an internal header in the Envoy configuration [...]

Then making "resource.name" work with `custom_tags` in Envoy is probably the best way to fix this. The operation name can be set to a literal elsewhere in the Envoy configuration, but that's just a literal. Adding a dedicated concept of "resource name" to Envoy is probably not worth it, especially since the term "resource" is used in OpenTelemetry and has a different meaning there than it does here.

I'll prioritize restoring the "resource.name" tag behavior as a bug fix.

**Comment by florianmutter** [comment]
I'm confused. We did update to Istio 1.20.0. Looking at [this](https://github.com/istio/proxy/blob/1.20.0/WORKSPACE#L37C8-L37C8) we do now use envoy 1.28. Looking at https://github.com/envoyproxy/envoy/blob/009bb79cc634fd621bfc2d84b6bd69e1da6a0046/source/extensions/tracers/datadog/tracer.cc#L94 this should contain the changed span name but we still don't see `envoy.proxy` being used as span name. Here is a screenshot:
Bildschirmfoto 2023-11-23 um 11 08 23
The same trace with Istio 1.17.5 that includes envoy 1.25 shows the same trace like this:
Bildschirmfoto 2023-11-23 um 11 18 11
Do you have any idea what that could be?

**Comment by njegosrailic** [comment]
Hi,

I believe the final patch was merged [last week](https://github.com/DataDog/dd-trace-cpp/issues/70#issuecomment-1812770804) for both envoy versions [1.27](https://github.com/envoyproxy/envoy/pull/30892) and [1.28](https://github.com/envoyproxy/envoy/pull/30893). These patches should be incorporated into the upcoming Envoy patch builds. Ideally, they will be released prior to the next Istio patch build to ensure their inclusion. I asked [yesterday](https://github.com/DataDog/dd-trace-cpp/issues/70#issuecomment-1820702332) whether this issue has been resolved, but it looks like my question only contributed to further confusion.

My main concern is that DataDog tracing remains _partially broken_ in both Istio versions `v1.18` and `v1.19`. This can be visually identified.

**Comment by florianmutter** [comment]
@njegosrailic Thank you very much. I was not aware of this.

**Comment by dgoffredo** [comment]
Thanks, @njegosrailic.

@florianmutter, it is surprising that the operation name of the span you screenshotted is not as expected.

One way to verify the versions of things is to look at the trace in all of its undecorated glory via the JSON API, i.e. `https://<datadog website>.com/api/v1/trace/3585287257395120597`. The last part of the path is the trace ID, which is displayed in the UI. The JSON dump of the trace (an array or spans) will contain the `_dd.tracer_version` tag, as well as a tag added by either Envoy or Istio telling the exact version of Envoy (I forget the name of the tag). This can be helpful for debugging.

In the case of the operation name bug, the precise Envoy version is what is important (the fix did not involve Datadog's core tracing library).

**Comment by njegosrailic** [comment]
Thanks, @dgoffredo.

Today, we tested the latest `v1.19` Istio patch, `v1.19.4`, and here are the important differences based on our research and the additional information you requested.

1. Trace name
```bash
<   "name": "test_app_target_edgegateway_http.test_app.svc.cluster.local_80",
---
>   "name": "envoy.proxy",
```

2. Trace resource
```bash
<   "resource": "egress istio-test-app-target-edgegateway-http-c3.service.testzone.internal",
<   "resource_hash": "423a0107c5aec597",
---
>   "resource": "test-app-target-edgegateway-http.test-app.svc.cluster.local:80/*",
>   "resource_hash": "647e5ccae10c6628",
```

3. Envoy version
```bash
<     "version": "envoy f9707e29aa0a36b1430b373d95e6c9abb5deca75/1.27.3-dev/Clean/RELEASE/BoringSSL",
---
>     "version": "envoy ea8f1b7c160f265a3fcfaab33446b8f5de98ca11/1.24.8-dev/Clean/RELEASE/BoringSSL",
```

4. Datadog tracer version
```bash
<     "_dd.p.dm": "-1",
<     "_dd.tracer_version": "v0.1.8",
---
>     "_dd.tracer_version": "v1.2.1",
```

**Comment by dgoffredo** [comment]
@njegosrailic That's very helpful.
```diff
<     "version": "envoy f9707e29aa0a36b1430b373d95e6c9abb5deca75/1.27.3-dev/Clean/RELEASE/BoringSSL",
---
>     "version": "envoy ea8f1b7c160f265a3fcfaab33446b8f5de98ca11/1.24.8-dev/Clean/RELEASE/BoringSSL",
```

New Istio's Envoy is on top, and old Istio's Envoy is on bottom.

The "version" names a git tree in [istio/proxy][1], not in the upstream Envoy. The tree named by the output of the newer Istio, 
`f9707e29aa0a36b1430b373d95e6c9abb5deca75` is [this][2]. Its [WORKSPACE][3] file names the _commit_ of upstream Envoy on which it's based, [c658fab86a174fda07961e196edb97f9da9b5a50][4]. That's the commit that fixes the sampling issue, backported onto Envoy's v1.27 release branch. Looking at the tree at that commit, I see that my "operation name" fix is _also_ in [that code][5].

So, the change to operation name shown elsewhere in your diff is unexpected.

Please convey this issue to Datadog Support. They will collect configuration information, and we can take it from there. 

[1]: https://github.com/istio/proxy
[2]: https://github.com/istio/proxy/tree/f9707e29aa0a36b1430b373d95e6c9abb5deca75
[3]: https://github.com/istio/proxy/blob/f9707e29aa0a36b1430b373d95e6c9abb5deca75/WORKSPACE#L38C14-L38C54
[4]: https://github.com/envoyproxy/envoy/commit/c658fab86a174fda07961e196edb97f9da9b5a50
[5]: https://github.com/envoyproxy/envoy/blob/c658fab86a174fda07961e196edb97f9da9b5a50/source/extensions/tracers/datadog/tracer.cc#L90-L94

**Comment by florianmutter** [comment]
I attached two spans of a similar trace. I removed all other spans.

The first span `6214739122545778382` is created by the istio gateway. Here the version is `envoy 580bf9cc561a012d4eb14af48d496d6748307e59/1.28.1-dev/Clean/RELEASE/BoringSSL` and the `_dd.tracer_version` is `0.1.8`.

The second span `1767944568465430022` is created by the istio sidecar of a service. The version is taken from the `DD_VERSION` variable so we do not see an envoy version. `_dd.tracer_version` is also `0.1.8`. When I check the side car it runs the same `proxyv2` image that is used in the gateway. So it is very very likely the same envoy version. We also only have one version of istio deployed.

Istio gateway span:
```json
{
  "trace_id": "6214739122545778382",
  "span_id": "6214739122545778382",
  "parent_id": "0",
  "start": 1701244677.028113,
  "end": 1701244677.184877,
  "duration": 0.156763803,
  "type": "web",
  "service": "istio-ingressgateway-iap.istio-system",
  "name": "gateway.instore.svc.cluster.local_8080_api",
  "resource": "egress instore.internal.example.com",
  "resource_hash": "f17c52a6c152467c",
  "host_id": 563956795,
  "hostname": "gke-development-pool3-41f15120-m6ns.europe-west4-a.c.puc-d-ngshop-gke-0228.internal",
  "env": "development",
  "host_groups": [
    "env:development"
  ],
  "meta": {
    "istio.mesh_id": "cluster.local",
    "version": "envoy 580bf9cc561a012d4eb14af48d496d6748307e59/1.28.1-dev/Clean/RELEASE/BoringSSL",
    "http.protocol": "HTTP/2",
    "http.status_code": "200",
    "runtime-id": "afd3478a-2e0a-413b-a8d1-74ae1360d99c",
    "response_flags": "-",
    "http.useragent": "axios/1.6.2",
    "zone": "europe-west4",
    "_dd.p.dm": "-1",
    "process_id": "14",
    "http.url_details.host": "instore.internal.example.com",
    "language": "cpp",
    "istio.namespace": "istio-system",
    "http.method": "POST",
    "upstream_cluster.name": "outbound|8080||gateway.instore.svc.cluster.local",
    "peer.address": "10.239.100.136",
    "istio.canonical_service": "istio-ingressgateway-iap",
    "_dd.agent_hostname": "gke-development-pool3-41f15120-m6ns.europe-west4-a.c.puc-d-ngshop-gke-0228.internal",
    "_dd.agent_version": "7.49.0",
    "http.url_details.path": "/api/servicepoint/stock-info/click-and-reserve",
    "http.url": "https://instore.internal.example.com/api/servicepoint/stock-info/click-and-reserve?tenant=PUC_DE",
    "http.path_group": "/api/servicepoint/stock-info/click-and-reserve",
    "request_size": "139",
    "http.host": "instore.internal.example.com",
    "node_id": "router~10.239.66.6~istio-ingressgateway-iap-867b5b4c79-mzht6.istio-system~istio-system.svc.cluster.local",
    "_dd.agent_rare_sampler.enabled": "false",
    "istio.canonical_revision": "latest",
    "_dd.tracer_version": "v0.1.8",
    "component": "proxy",
    "http.url_details.scheme": "https",
    "downstream_cluster": "-",
    "guid:x-request-id": "52d096ff-0b0c-9df6-a33e-fbd156869bf4",
    "ddtags": "ingestion_reason:auto",
    "http.url_details.queryString.tenant": "PUC_DE"
  },
  "metrics": {
    "_dd.agent_errors_sampler.target_tps": 10.0,
    "_dd.agent_psr": 1.0,
    "_trace_root": 1.0,
    "response_size": 74.0,
    "_sampling_priority_v1": 1.0,
    "_top_level": 1.0,
    "_dd.agent_priority_sampler.target_tps": 10.0
  },
  "ingestion_reason": "auto",
  "metadata": {
    "sds_info": []
  },
  "children_ids": [
    "2310811855767820061"
  ]
}
```

Istio sidecar span
```json
{
  "trace_id": "6214739122545778382",
  "span_id": "1767944568465430022",
  "parent_id": "5261554865300988373",
  "start": 1701244677.057804,
  "end": 1701244677.179114,
  "duration": 0.121309735,
  "type": "web",
  "service": "fulfill.order",
  "name": "order.fulfill.svc.cluster.local_50051",
  "resource": "ingress",
  "resource_hash": "9498ccf48d93a1bf",
  "host_id": 563547440,
  "hostname": "gke-development-pool3-cd9ee151-7dsz.europe-west4-c.c.puc-d-ngshop-gke-0228.internal",
  "env": "development",
  "host_groups": [
    "env:development"
  ],
  "meta": {
    "upstream_cluster.name": "inbound|50051||",
    "zone": "europe-west4",
    "http.status_code": "200",
    "downstream_cluster": "-",
    "http.useragent": "grpc-java-netty/1.59.0",
    "istio.mesh_id": "cluster.local",
    "http.url_details.host": "order.fulfill",
    "language": "cpp",
    "http.url_details.scheme": "http",
    "_dd.agent_version": "7.49.0",
    "_dd.tracer_version": "v0.1.8",
    "component": "proxy",
    "istio.namespace": "fulfill",
    "http.url_details.path": "/fulfill.order.OrderService/CreateNewRetailOrder",
    "grpc.authority": "order.fulfill:50051",
    "grpc.path": "/fulfill.order.OrderService/CreateNewRetailOrder",
    "version": "rev-d0eb43b93f93612ce20b31bb7ce4e45d83975145",
    "guid:x-request-id": "b10e2e1f-4d3b-939e-bd6a-974d35c31184",
    "istio.canonical_service": "order",
    "node_id": "sidecar~10.239.67.58~order-6cdf99b645-lb582.fulfill~fulfill.svc.cluster.local",
    "grpc.status_code": "0",
    "peer.address": "10.239.67.87",
    "grpc.content_type": "application/grpc",
    "http.protocol": "HTTP/2",
    "http.url": "http://order.fulfill:50051/fulfill.order.OrderService/CreateNewRetailOrder",
    "ddtags": "ingestion_reason:auto",
    "_dd.p.dm": "-1",
    "env": "development",
    "http.url_details.port": "50051",
    "runtime-id": "5193c159-e218-44f1-8750-34d1d454fcd9",
    "http.path_group": "/fulfill.order.OrderService/CreateNewRetailOrder",
    "response_flags": "-",
    "request_size": "347",
    "_dd.agent_hostname": "gke-development-pool3-cd9ee151-7dsz.europe-west4-c.c.puc-d-ngshop-gke-0228.internal",
    "istio.canonical_revision": "latest",
    "process_id": "23",
    "_dd.agent_rare_sampler.enabled": "false",
    "http.host": "order.fulfill",
    "http.method": "POST"
  },
  "metrics": {
    "_top_level": 1.0,
    "_dd.agent_priority_sampler.target_tps": 10.0,
    "response_size": 39.0,
    "_dd.agent_errors_sampler.target_tps": 10.0,
    "_sampling_priority_v1": 1.0
  },
  "ingestion_reason": "auto",
  "metadata": {
    "sds_info": []
  },
  "children_ids": [
    "6764890222525977566"
  ]
}
```

**Pull request merged**